routing/index_router.cpp

```diff
@@ -205,37 +205,43 @@ void PushPassedSubroutes(Checkpoints const & checkpoints, vector<Route::Subroute
 namespace routing
 {
 // IndexRouter::BestEdgeComparator ----------------------------------------------------------------
-IndexRouter::BestEdgeComparator::BestEdgeComparator(m2::PointD const & point, m2::PointD const & direction)
+IndexRouter::BestEdgeComparator::BestEdgeComparator(m2::PointD const & point,
+                                                    m2::PointD const & direction)
   : m_point(point), m_direction(direction)
 {
 }
 
-bool IndexRouter::BestEdgeComparator::Compare(Edge const & edge1, Edge const & edge2) const
+int IndexRouter::BestEdgeComparator::Compare(Edge const & edge1, Edge const & edge2) const
 {
-  if (!m_direction.IsAlmostZero())
+  if (IsDirectionValid())
   {
     bool const isEdge1Parallel = IsAlmostParallel(edge1);
     if (isEdge1Parallel != IsAlmostParallel(edge2))
-      return isEdge1Parallel;
+      return isEdge1Parallel ? -1 : 1;
   }
 
-  return GetSquaredDist(edge1) < GetSquaredDist(edge2);
+  double const squaredDistFromEdge1 = GetSquaredDist(edge1);
+  double const squaredDistFromEdge2 = GetSquaredDist(edge2);
+  if (squaredDistFromEdge1 == squaredDistFromEdge2)
+    return 0;
+
+  return squaredDistFromEdge1 < squaredDistFromEdge2 ? -1 : 1;
 }
 
 bool IndexRouter::BestEdgeComparator::IsAlmostParallel(Edge const & edge) const
 {
-  CHECK(!m_direction.IsAlmostZero(), ());
+  CHECK(IsDirectionValid(), ());
 
   m2::PointD const edgeDirection =
-      edge.GetEndJunction().GetPoint() - edge.GetStartJunction().GetPoint();
+    edge.IsForward() ? edge.GetEndJunction().GetPoint() - edge.GetStartJunction().GetPoint()
+                     : edge.GetStartJunction().GetPoint() - edge.GetEndJunction().GetPoint();
   if (edgeDirection.IsAlmostZero())
     return false;
 
-  double const cosAng =
-    m2::DotProduct(m_direction.Normalize(), edgeDirection.Normalize());
+  double const cosAng = m2::DotProduct(m_direction.Normalize(), edgeDirection.Normalize());
   // Note. acos(0.97) ≈ 14 degrees.
   double constexpr kMinCosAngForAlmostParallelVectors = 0.97;
-  return  cosAng >= kMinCosAngForAlmostParallelVectors;
+  return cosAng >= kMinCosAngForAlmostParallelVectors;
```

Comment:
https://ru.wikipedia.org/wiki/%D0%9A%D0%BE%D0%BB%D0%BB%D0%B8%D0%BD%D0%B5%D0%B0%D1%80%D0%BD%D0%BE%D1%81%D1%82%D1%8C